allennlp/training/optimizers.py

```diff
@@ -196,6 +205,159 @@ def default(model_parameters: List) -> "Optimizer":
+        
+        super().__init__(parameter_groups, optimizer_groups["default"][1])
+
+    def _populate_optimizer_groups(self, parameter_groups):
+        """
+        Assigns each parameter group in `parameter_groups` to the
+        appropriate optimizer or to the 'default' optimizer.
+        """
+        optimizer_groups = {}
+
+        for parameter_group in parameter_groups:
+            # Check to see what optimizer this group should be assigned to.
+            if "name" in list(parameter_group.keys()):
+                optimizer_key = parameter_group["name"]
+                if not optimizer_key in optimizer_groups:
+                    optimizer_groups[optimizer_key] = ([], {})
+                
+                group = {key: value for key, value in parameter_group.items() if key != "name"}
+                optimizer_groups[optimizer_key][0].append(group)
+            # If no optimizer name is given, assign this group to the default group.
+            else:
+                if not "default" in optimizer_groups:
+                    optimizer_groups["default"] = ([], {})
+                group = {key: value for key, value in parameter_group.items()}            
+                optimizer_groups["default"][0].append(group)
+        
+        return optimizer_groups
+
+    def _set_optimizer_kwargs(self, optimizer_groups, optimizers):
+        """
+        Sets the optimizer keyword arguments for each optimizer.
+        """
+        for optimizer in optimizers:
+            optimizer_key = optimizer["name"]
+            for key, value in optimizer.items():
+                if key != "name":
+                    # Set optimizer kwargs.
+                    optimizer_groups[optimizer_key][1][key] = value
+
+    @overrides
+    def step(self):
+        """
+        Takes an optimization step for each optimizer.
+        """
+        for optimizer in self._grouped_optimizers.values():
+            optimizer.step()
+
+    @overrides
+    def state_dict(self):
+        """
+        Creates an object `optimizer_state_dict`, which is a dictionary mapping an optimizer key to its `state_dict`.
+        This dictionary is used as the value for 'optimizer' in the 'training_states' dictionary in the `gradient_descent`
+        `Trainer`, e.g. "optimizer" : {"optimizer1": `optimizer1_state_dict`, "optimizer2" : `optimizer2_state_dict`}.
+        """
+        optimizer_state_dict = {
+            f"{optimizer_key}_optimizer": optimizer.state_dict() for optimizer_key, optimizer in self._grouped_optimizers.items()
+        }
+        
+        return optimizer_state_dict
+
+    @overrides
+    def load_state_dict(self, training_state: Dict[str, Any]):
+        """
+        Loads each optimizer's `state_dict`.
+        """
+        for optimizer_key, optimizer in self._grouped_optimizers.items():
+            optimizer.load_state_dict(training_state[f"{optimizer_key}_optimizer"])
+
+    @overrides
+    def zero_grad(self, set_to_none: bool = False):
```

Comment:
`set_to_none` enables setting gradients to `None` or zero, see [here for more info](https://pytorch.org/docs/stable/optim.html#torch.optim.Optimizer.zero_grad). I don't like hard-coding the Boolean here but perhaps I could add a parameter `set_gradients_to_none` to `RegexOptimizer` to change this. But the `Trainer` doesn't call `optimizer.zero_grad()` so I didn't implement this.

Reply:
What's hardcoded? Just the default value?

Reply:
Yes, just the default value of `set_to_none`, i.e. there is nowhere in the code where you can change this to `True`

Reply:
We don't usually count default parameters as "hardcoded". They are just defaults.

Also, they become part of the function signature. Python will let you change them, but you really shouldn't for maximum compatibility.